drivers/wireless/bluetooth/bt_rpmsghci.c

```diff
@@ -0,0 +1,535 @@
+ *   rdev  - The rpmsg-hci end point
+ *   priv_ - Rpmsg-hci handle
+ *
+ * Returned Values:
+ *   None
+ *
+ ****************************************************************************/
+
+static void bt_rpmsghci_client_created(FAR struct rpmsg_device *rdev,
+                                       FAR void *priv_)
+{
+  FAR struct rpmsghci_s *priv = priv_;
+  char                   eptname[RPMSG_NAME_SIZE];
+
+  if (strcmp(priv->cpuname, rpmsg_get_cpuname(rdev)) == 0)
+    {
+      snprintf(eptname, RPMSG_NAME_SIZE, "%s%s", RPMSGHCI_NAME_PREFIX,
+               priv->name);
+      priv->ept.priv = priv;
+      rpmsg_create_ept(&priv->ept, rdev, eptname,
+                       RPMSG_ADDR_ANY, RPMSG_ADDR_ANY,
+                       bt_rpmsghci_client_ept_cb, NULL);
+    }
+}
+
+/****************************************************************************
+ * Name: bt_rpmsghci_client_destroy
+ *
+ * Description:
+ *   Rpmsg HCI destroy function, this function will be called by rptun to
+ *   destroy rpmsg-HCI end point.
+ *
+ * Parameters:
+ *   rdev  - The rpmsg-HCI end point
+ *   priv_ - Rpmsg-HCI handle
+ *
+ * Returned Values:
+ *   None
+ *
+ ****************************************************************************/
+
+static void bt_rpmsghci_client_destroy(FAR struct rpmsg_device *rdev,
+                                       FAR void *priv_)
+{
+  FAR struct rpmsghci_s *priv = priv_;
+
+  if (!strcmp(priv->cpuname, rpmsg_get_cpuname(rdev)))
+    {
+      rpmsg_destroy_ept(&priv->ept);
+    }
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: rpmsghci_register
+ *
+ * Description:
+ *   Rpmsg-HCI client initialize function, the client cpu should call
+ *   this function in the board initialize process.
+ *
+ * Parameters:
+ *   cpuname  - the server cpu name
+ *   name     - the HCI device you want to access in the remote cpu
+ *
+ * Returned Values:
+ *    A non-NULL handle is returned on success;
+ *    A NULL is returned on any failure.
+ *
+ ****************************************************************************/
+
+FAR struct bt_driver_s *rpmsghci_register(FAR const char *cpuname,
+                                          FAR const char *name)
+{
+  FAR struct rpmsghci_s *priv = NULL;
+
+  /* Validate input */
+
+  if (!cpuname || !name)
+    {
+      goto errout;
+    }
+
+  /* Create RPMSG-HCI driver */
+
+  priv = kmm_zalloc(sizeof(struct rpmsghci_s));
+  if (!priv)
+    {
+      goto errout;
+    }
+
+  /* Initialize BT driver */
+
+  priv->btdev.head_reserve = 0;
+  priv->btdev.open         = bt_rpmsghci_open;
+  priv->btdev.close        = bt_rpmsghci_close;
+  priv->btdev.send         = bt_rpmsghci_send;
+  priv->btdev.ioctl        = bt_rpmsghci_ioctl;
+  priv->btdev.rpmsg        = priv;
```

Comment:
why not save to btdev.piv

Reply:
btdev.priv is used by the BTH4 driver, so this won't work if the user want register HCI controller as BTH4

Reply:
done